### PR TITLE
[Issue 364] Stop Consolidation of Warnings

### LINKED ISSUE #364: Show all warnings
### OpenTF Version

```shell
1.0
```

### Use Cases

When upgrading a provider eg the aws provider, many warnings can occur, but only the first of them is shown. Eg 
```
Warning: Argument is deprecated
  on .terraform/modules/spa.s3_logs/main.tf line 1, in resource "aws_s3_bucket" "this":
   1: resource "aws_s3_bucket" "this" {
Use the aws_s3_bucket_acl resource instead

(and 21 more similar warnings elsewhere)
```
There is currently no option to show those similar warnings. All these warnings could be in different files, so we are forced to first address the only one shown, then re-run the terraform plan. Sometimes this requires many cycles. TEDIOUS!

### Attempted Solutions

Run `terraform validate`. 

However this does not seem to catch these warnings, possibly because the warnings are not about invalid code, just deprecation warnings. 

### Proposal

Several possibilities: 
- Add a switch on the plan and apply commands, eg `--show-all-warnings`. This switch should work even if the output is json. 
- Add a new command, `show-warnings`: if the users runs an apply or a plan and sees deprecation warnings, then they could run this new command to get the details. This makes sense if there are steps of the plan / apply can be skipped when the user just wants these (eg one would expect that it would be sufficient for terraform to parse the code, and ask the provider for any warnings about it, without actually having to check the cloud resources status, which could save a ton of time)

### References

- The issue is mentioned in a few places but the biggest discussion I found is https://discuss.hashicorp.com/t/show-all-warnings/5606/18. This demonstrates that HC is not interested in implementing this much-needed and almost trivial capability.

### PR DESCRIPTION
**Pull request by Hefeweizen**
Change:
add a flag to show all instances of similar warnings

Rationale:
Multiple instances of similar warnings are consolidated into one entry, with notation that there may be more.  This obscures the location of those additional warnings.  If one desires to change those warnings, one must iteratively fix them, rerunning the plan after each change, to get the tooling to show you these changes.

This new flag stops the consolidation.

Issue: https://github.com/opentffoundation/opentf/issues/364
Resolves #364

## Target Release

1.6.0

## Draft CHANGELOG entry

Add `-verbose-warnings` flag, optionally stopping the consolidation of similar warnings, instead listing the full details for every warning individually.

### NEW FEATURES | UPGRADE NOTES | ENHANCEMENTS | BUG FIXES | EXPERIMENTS

-  `opentf plan -verbose-warnings`: show details on all warnings.

### Example Output
```
$ opentf plan 
...
╷
│ Warning: Argument is deprecated
│ 
│   with aws_s3_bucket.b1,
│   on main.tf line 14, in resource "aws_s3_bucket" "b1":
│   14: resource "aws_s3_bucket" "b1" {
│ 
│ Use the aws_s3_bucket_versioning resource instead
│ 
│ (and 3 more similar warnings elsewhere)
╵
$ opentf plan -verbose-warnings
...
╷
│ Warning: Argument is deprecated
│ 
│   with aws_s3_bucket.b1,
│   on main.tf line 14, in resource "aws_s3_bucket" "b1":
│   14: resource "aws_s3_bucket" "b1" {
│ 
│ Use the aws_s3_bucket_versioning resource instead
╵
╷
│ Warning: Argument is deprecated
│ 
│   with aws_s3_bucket.b1,
│   on main.tf line 16, in resource "aws_s3_bucket" "b1":
│   16:   acl    = "private"
│ 
│ Use the aws_s3_bucket_acl resource instead
╵
╷
│ Warning: Argument is deprecated
│ 
│   with aws_s3_bucket.b2,
│   on main.tf line 23, in resource "aws_s3_bucket" "b2":
│   23: resource "aws_s3_bucket" "b2" {
│ 
│ Use the aws_s3_bucket_versioning resource instead
╵
╷
│ Warning: Argument is deprecated
│ 
│   with aws_s3_bucket.b2,
│   on main.tf line 25, in resource "aws_s3_bucket" "b2":
│   25:   acl    = "private"
│ 
│ Use the aws_s3_bucket_acl resource instead
╵
```

### CODE REVIEW

**Comment by Hefeweizen** [comment]
I'm eager for pointers and feedback.  I don't have any tests around the flag itself, I'll look into adding those.

**Comment by cube2222** [comment]
Hey @Hefeweizen, I appreciate the willingness to contribute, but I recommend waiting a bit (esp. with things like tests) until an issue is actually officially accepted to avoid wasted work.

**Comment by marcinwyszynski** [comment]
I know that there's been some discussions internally about allowing logs to be simultaneously written to stdout (human-readable) and a file in JSONL format. If that's implemented, would it make sense to leave the human-friendly output as-is and consume the machine-readable file for a (potentially very long) full list of warnings?

WDYT @Hefeweizen ?

**Comment by Hefeweizen** [comment]
I'll ask @schollii to weigh in on his thoughts; he had the original request.

My opinion?  Do both.  : )  Even if we have a jsonl logfile record to review at our leisure, some people will want the written output (maybe they don't like parsing jsonl with jq, or other tools).

If the concern is that the list of warnings gets too long, this PR could very easily be extended to introduce a cli flag of `-warning_count`.  Internally, we effectively set warning count to -1 (unlimited); it'd be straight-forward to accept a count from the user on how much they wanted to see.

And then there's a discussion around implementation.  The tool currently passes `diag` records back and forth.  This collected list of diags is checked for errors on many occasions but for user output, it's processed once[^1].  If we're wanting dual output (screen + jsonl), do you just process the diags twice?  That's easy enough, but if the command dies you have no output.  For more faithful logging, it seems like you'd want the jsonl logging inline where each warning/error is generated -- though this is a much more involved change.

In addition to getting errors as they're generated, you'd also see glimpses into the order of the processing of the DAG, which, in rare circumstances, may be interesting.  Though, I suddenly realize I'm also assuming single-threaded work; is that a valid assumption?  Ok, I'm rambling.  Ignore this last paragraph, but the rest of the response?  It's good.  : )



[^1]: diags are only processed once in any single command execution, but there's two locations where that can happen ([here](https://github.com/opentofu/opentofu/blob/1abb1073cdff03791d553d6f3009a94c304b4dba/internal/command/meta.go#L691C16-L691C31) and [here](https://github.com/opentofu/opentofu/blob/1abb1073cdff03791d553d6f3009a94c304b4dba/internal/command/views/view.go#L83)).  The intent was to move all commands to the view handling, but that has not been completed.  [Issue #420](https://github.com/opentofu/opentofu/issues/420)

**Comment by schollii** [comment]
I appreciate being flagged in for this. 

The JSONL output, is this when the `-json` option is provided on command line? Based on the above thread, it seems like it should be a separate concern, ie the human-consumable output should be human-friendly and the machine output (JSONL) should have everything always since it is amenable to further filtering sorting etc using jq and such. In which case the design implemented by current PR is fine. 

Also I think the current default of showing a summary is good, because sometimes there can be a 100 similar warnings that are due to one module being used in many places, and fixing it fixes the 100 :) 

One note of caution though: we have not discussed the alternative approach of a command: 
- As indicated in the issue I raised, long-term it might be better to implement a command like `opentf warnings <action> <args...>` or `opentf plan warnings <action> <args...>` which the user would run after seeing the plan show the warnings summary and wanting more info to determine which warning to fix first. Some actions would be easier via a command, like getting the details of only one of the groups of warnings. 
- warning systems rarely get simpler, ie the number and classes of warnings get larger and more sophisticated and varied such that more control of the warnings in a plan is inevitable, over time. 
 
That being said, I think the current PR's approach does not preclude a separate command eventually, from that point of view no further discussion is needed about alternatives, at this point, from my POV.

**Review by Yantrio — APPROVED** [review]

**Comment by Yantrio** [comment]
My approval is for the current code that is here now, I propose that we ensure this works e2e with the json and human output nicely before merging.

**Review by cam72cam — APPROVED** [review]
Code approved, with the caveat that we need a changelog entry committed (Core team could do this after merge if needed).

Also I'd like to make sure `-verbose-warnings` is what we want for the flag name.  `-consolidate-warnings = false` could also work.  Verbose and Non Compact are overlapping but not identical in my mind.

Edited from compact -> consolidate

**Comment by cam72cam** [comment]
@Hefeweizen would you be able to rebase this PR against main?  We have added a lot of new workflows that github is not happy running on this older branch.

**Comment by cam72cam** [comment]
Rebased and extended in https://github.com/opentofu/opentofu/pull/1894.  Thank you for your initial work on this @Hefeweizen!

**Pull request closed without merging**
